# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmIonCompile.cpp
# Commit: cfe5dc930b4c
# Full Hash: cfe5dc930b4c234f60d064abac3cb052c98ab65e
# Author: Lars T Hansen <lhansen@mozilla.com>
# Date: 2020-06-03 04:20:58
# Description:
#   Bug 1641973 - Make sure to check inDeadCode. r=bbouvier
#   
#   Several of the SIMD instruction emitters in WasmIonCompile did not
#   check the dead-code flag, and some might therefore access null pointers
#   resulting from popping the value stack in dead code regions.
# ==============================================================================

diff -r 400db4104b94 -r cfe5dc930b4c js/src/wasm/WasmIonCompile.cpp
--- a/js/src/wasm/WasmIonCompile.cpp	Tue Jun 02 09:47:15 2020 +0000
+++ b/js/src/wasm/WasmIonCompile.cpp	Tue Jun 02 09:58:25 2020 +0000
@@ -730,6 +730,10 @@
   // (v128, v128, v128) -> v128 effect-free operations
   MDefinition* bitselectSimd128(MDefinition* v1, MDefinition* v2,
                                 MDefinition* control) {
+    if (inDeadCode()) {
+      return nullptr;
+    }
+
     MOZ_ASSERT(v1->type() == MIRType::Simd128);
     MOZ_ASSERT(v2->type() == MIRType::Simd128);
     MOZ_ASSERT(control->type() == MIRType::Simd128);
@@ -740,6 +744,10 @@
 
   // (v128, v128, imm_v128) -> v128 effect-free operations
   MDefinition* shuffleSimd128(MDefinition* v1, MDefinition* v2, V128 control) {
+    if (inDeadCode()) {
+      return nullptr;
+    }
+
     MOZ_ASSERT(v1->type() == MIRType::Simd128);
     MOZ_ASSERT(v2->type() == MIRType::Simd128);
     auto* ins = MWasmShuffleSimd128::New(
@@ -752,6 +760,10 @@
   MDefinition* loadSplatSimd128(Scalar::Type viewType,
                                 const LinearMemoryAddress<MDefinition*>& addr,
                                 wasm::SimdOp splatOp) {
+    if (inDeadCode()) {
+      return nullptr;
+    }
+
     // Expand load-and-splat as integer load followed by splat.
     MemoryAccessDesc access(viewType, addr.align, addr.offset,
                             bytecodeIfNotAsmJS());
@@ -766,6 +778,10 @@
 
   MDefinition* loadExtendSimd128(const LinearMemoryAddress<MDefinition*>& addr,
                                  wasm::SimdOp op) {
+    if (inDeadCode()) {
+      return nullptr;
+    }
+
     MemoryAccessDesc access(Scalar::Int64, addr.align, addr.offset,
                             bytecodeIfNotAsmJS());
     // Expand load-and-extend as integer load followed by widen.
